### ISSUE TITLE
When should app history state be carried over?

### ISSUE THREAD

**Issue by domenic** [WICG/app-history#100]
The current state of the explainer is:

The following carries over app history state to the next entry:

- Fragment navigation, including `appHistory.navigate("#foo")` and `location.hash = "foo"`

The following reload the current entry and so leave the app history state as-is:

- `appHistory.navigate({ navigateInfo })`
- `location.reload()`

The following replace current entry with a new entry whose the app history state is undefined:

- `history.replaceState()`
- `appHistory.navigate(appHistory.current.url)`
- `appHistory.navigate(appHistory.current.url, { replace: true })` (here `replace` is redundant)
- `appHistory.navigate(otherURL, { replace: true })`
- `location.href = location.href`

(Note: in general same-URL navigations are not interoperable for `history.state`: https://github.com/whatwg/html/issues/6213 . We need to pick a behavior and test it to death for app history. Currently we've picked reset to undefined.)

The following cases push a new entry and do not carry over the app history state, so the new state will be undefined:

- `history.pushState()`
- `appHistory.navigate(otherURL)`
- `location.href = otherURL`

The following, of course, explicitly set the app history state to the new value:

- `appHistory.navigate(url, { state })`
- `appHistory.navigate(url, { state, replace: true })`
- `appHistory.navigate({ state })`

---

This doesn't seem very principled. I think what happened here is that we heard "it's annoying that `history.state` gets reset on fragment navigations", and fixed that. But then we kept doing the natural-from-the-spec's-point-of-view thing and reset by default.

I think I have two main questions:

1. Should replaces carry over the state by default? (I.e., always carry over, unless you do `appHistory.navigate(url, { state, replace: true })` with a new `state` value.)

2. Should we carry over the state on any of the same-document push cases? Recall that `appHistory.navigate(otherURL)` and `location.href = otherURL` can all be turned into same-document navigations via the `navigate` event.

**Comment by bathos** [comment]
We experimented previously with carrying over state by default in all cases. It seemed like a good idea initially but I regretted it pretty quickly.

The trouble was that most state which made sense to carry forward automatically turned out not to _really_ belong to the state of any history entry at all. Although I suspect there really are cases where it would make sense, the automatic carrying-forward encouraged usage of the history state as if it were interchangeable with sessionStorage, and because folks navigate forwards a lot more often than backwards, it was easy to miss when history entry state was not the appropriately storage mechanism for something.

We’re still doing automatic carry over on replacement, but I plan to get rid of this too; the problem seems no different there. The exception is the hashchange case, where carryover is definitely helpful and has helped fix problems (rather than creating new ones).
